### DISCUSSION TITLE
Pens: automatic processing

### DISCUSSION THREAD

**Discussion by madig** [fonttools/fonttools#2204] [General]
So, while fiddling with https://github.com/fonttools/fonttools/pull/2202, I found that there's some magic in adapter pens like BasePointToSegmentPen that drops empty contours on endPath, shifts contours so they start on an on-curve and turns single-point contours into open ones. I am not familiar with the complexities of adapters so I'm wondering if these are optimizations or rather intrinsically needed? Should point pens be expected to do any of these optimizations, too?

Side question: why does ReverseContourPointPen require that a component is only added while a contour drawing is not in progress?

**Comment by justvanrossum** [comment]
> drops empty contours on endPath 

Empty contours can't be represented with segment pens

> shifts contours so they start on an on-curve

Segment pens can't start a path on an offcurve (except for the on-curve-less quad blob)

> turns single-point contours into open ones

Not sure, but a closed single-point path doesn't make much sense, either

> why does ReverseContourPointPen require that a component is only added while a contour drawing is not in progress

Because it doesn't make sense to add a component in between a beginPath and an endPath call. Perhaps that's an aspect of the protocol that should be formally described. There's no real technical reason to disallow it, it just doesn't make sense: why would one want to do that?

**Reply by madig** [reply]
Thanks! That makes sense. I shall turn these into code comments... one day.

> Because it doesn't make sense to add a component in between a beginPath and an endPath call

Hm. I suppose I can see the potential confusion when something decomposes a component on arrival while a path is in progress, but it's the first time I have stumbled over the restriction. I can add this as a requirement to the code though. Or maybe it should not be blocked unless the pen needs it? Dunno...

**Reply by justvanrossum** [reply]
A component is always a full path, it doesn't make sense to me to allow firstHalfPath + fullPath + secondHalfPath. This goes both for point pens and segment pens: adding a component between moveTo and closePath/endPath is equally problematic. Just in case you ever get to write those comments :)

**Reply by justvanrossum** [reply]
In other words: a component (in the pen world) is not an arbitrary CFF-style path subroutine.

**Reply by madig** [reply]
Thanks! That clears some stuff up. Another question I forgot: the PointToSegment pen will rotate the point list so that the first on-curve point becomes the last point in the point list. Is this a normalization technique to ensure that you don't have to wrap around the point list mid-flight?

**Reply by justvanrossum** [reply]
I must admit that the code is a bit hard to follow, but I bet it's not about normalization, but simply the implementation strategy to ensure this:

> Segment pens can't start a path on an offcurve (except for the on-curve-less quad blob)

**Reply by justvanrossum** [reply]
Symmetry with SegmentToPointPen may also be a factor.

**Reply by madig** [reply]
Ok, thanks :)

**Reply by madig** [reply]
I *have* to loop back on this. I kept wondering just what the problem with starting on an off-curve is, all the segment pen methods take sequences of points, which for curves should start with off-curves IIUC. So @cmyr offered this explanation:

> Colin Rofls: so I've had to fight this, and I can offer my understanding. 
>
> In an open path, UFO requires the first point in the list to be an on-curve, which is the "moveTo".
>
> in a closed path, UFO only says that there is no explicit start point; the points are treated as a circular buffer, and any order is acceptable.
>
> However: we still need somewhere to actually "moveTo" when we start drawing. If the first points are off-curves, and we decide to treat, say, the 3rd point as the start point, at the end of all the points we need to loop back and handle that point again, when we close the path.
>
> We can avoid this book-keeping if the "moveTo" is also the last point in the array; we do the move, and then we just iterate through all the points, and when we're done we're back where we started.
>
> This has another nice property, for implementors: when you start a path it is always open; but when you close the path, you don't actually add a point; you just mark the path as closed, and rotate the points to the left, putting the previous start at the end, giving you the order described above.
>
> Basically I would say that this "in a closed path, the last point should be an on-curve, and is the start point" convention has a certain simple elegance, and I frankly think it should be part of the spec.
>
> (there may be a good reason why you wouldn't want it, but I don't know what that is.)
>
> ... did that answer your question?
>
> me: so by putting the starting on-curve last, we also move its potential offcurves last (i.e. if we have a circle and start on the curve and the contour list ends on two offcurves, rotating means we don't have to loop back because all we need is in the last motion of the pen
>
> Colin Rofls: right, it means we don't ever need any looping logic.

**Reply by cmyr** [reply]
If I knew this was going to be shared past zulip I would've gone back and edited out all those semicolons. 🤕

**Reply by madig** [reply]
Sorry :speak_no_evil: I think the semicolons are fine

**Reply by justvanrossum** [reply]
I've totally lost you. What is the problem with fonttools point pens here?

**Reply by madig** [reply]
There is no problem, I just wanted to better understand why the starting point is moved to the end.

**Reply by madig** [reply]
More questions!

```
In [5]: p = _TestPen()
In [6]: p.qCurveTo((100,100))
100 100 lineto
```
```
In [9]: r2 = RecordingPen()
In [10]: p2 = PointToSegmentPen(r2)
In [12]: p2.beginPath()
In [13]: p2.addPoint((100,100), segmentType="qcurve")
In [14]: p2.endPath()
In [15]: r2.value
Out[15]: [('moveTo', ((100, 100),)), ('endPath', ())]
```

BasePen turns single points into lines, PointToSegmentPen into moves?

**Reply by justvanrossum** [reply]
You're comparing apples and oranges.

- A segment pen subpath *must* start with a moveTo, unless you're drawing a quad blob (last qCurve arg is None), so the first example is invalid.

- qCurve takes one or more arguments. The one-argument case is debatable (perhaps it shouldn't have been allowed), but a single on-curve point does not make a curve. BasePen can't call _qCurveOne, so it falls back to a line. (It's based on TT curves: two consecutive on-curve points form a line.)

- You're converting a point pen sequence to segments. You're adding a single on-curve point, but nothing else. This is the same as adding a single "line" point, and PointToSegmentPen turns a single-point closed contour into a single-point open contour, for whatever reason. But it _has_ to output a moveTo to the receiving segment pen, because that's how segment pens work (see the first bullet).

**Reply by madig** [reply]
Thanks but :exploding_head: 

> A segment pen subpath must start with a moveTo, unless you're drawing a quad blob (last qCurve arg is None), so the first example is invalid.

Why is the quad blob exempt?

**Reply by justvanrossum** [reply]
Because a quad blob has no on-curve points at all. Otherwise it would not be a quad blob. It's a TT-specific thing that fits poorly in the segment pen protocol, but I felt was important enough to create a weird exception for.
